### ISSUE TITLE
Ultimate Gear achievement badge is not being awarded for users with no false value for that badge

### ISSUE THREAD

**Issue by Alys** [HabitRPG/habitica#5427]
We've had a couple of reports of players buying the final gear for their class but not being given the Ultimate Gear achievement (I've awarded their achievements manually). I've done some testing and found the cause:

The `achievements.ultimateGearSets` object in a player's account can look like either one of these (note the `false` values in the first one):

```
"achievements.ultimateGearSets" : {
        "healer"  : false,
        "rogue"   : false,
        "warrior" : true,
        "wizard"  : false
    }
```

```
"achievements.ultimateGearSets" : {
        "warrrior"  : true,
    }
```

If the user has the first object, they WILL get new achievements. If they have the second object, they will NOT. This should affect only older users. New ones created since the Ultimate Gear code was last tweaked should be okay.

I will fix this by doing a migration that gives correct true AND false values to all users who need them, so this bug will be fixed in a few hours (migration is slooooowwwww and I have a few other things to do first).

But, we might also want the code to be changed so that it could handle a missing false value anyway. It seems like a correct, robust thing to do. @crookedneighbor @gisikw What do you think? Or it is just pointless work once we have correct data for all existing users? It bothers me to rely on the data though. It's a fix that a mid-level contributor could do, or a new one with experience in coffeescript.

The relevant code is in `common/script/index.coffee`, starting at around line 1713. At a guess, `if user.achievements.ultimateGearSets[klass] is not true` might be failing if the value is non-existent.

To assist with testing, should it be necessary, below are the item keys for the ultimate gear in each class, an update command that will give all gear EXCEPT the ultimate pieces to a test user (and also level, gold, and gems for buying gear and class changes), and a find command to display appropriate data.

```
armor_healer_5
head_healer_5
shield_healer_5
weapon_healer_6

armor_rogue_5
head_rogue_5
shield_rogue_6
weapon_rogue_6

armor_warrior_5
head_warrior_5
shield_warrior_5
weapon_warrior_6

armor_wizard_5
head_wizard_5
weapon_wizard_6
```

```
db.users.update({'_id':uuid}, {$set: { 
'stats.lvl':20, 'stats.gp':10000, "balance":10,
"items.gear.owned": {
    "armor_healer_1": true, "armor_healer_2": true, "armor_healer_3": true, "armor_healer_4": true,
    "armor_rogue_1": true, "armor_rogue_2": true, "armor_rogue_3": true, "armor_rogue_4": true,
    "armor_warrior_1": true, "armor_warrior_2": true, "armor_warrior_3": true, "armor_warrior_4": true,
    "armor_wizard_1": true, "armor_wizard_2": true, "armor_wizard_3": true, "armor_wizard_4": true,
    "head_healer_1": true, "head_healer_2": true, "head_healer_3": true, "head_healer_4": true,
    "head_rogue_1": true, "head_rogue_2": true, "head_rogue_3": true, "head_rogue_4": true,
    "head_warrior_1": true, "head_warrior_2": true, "head_warrior_3": true, "head_warrior_4": true,
    "head_wizard_1": true, "head_wizard_2": true, "head_wizard_3": true, "head_wizard_4": true,
    "shield_healer_1": true, "shield_healer_2": true, "shield_healer_3": true, "shield_healer_4": true,
    "shield_rogue_0": true, "shield_rogue_1": true, "shield_rogue_2": true, "shield_rogue_3": true, "shield_rogue_4": true, "shield_rogue_5": true,
    "shield_warrior_1": true, "shield_warrior_2": true, "shield_warrior_3": true, "shield_warrior_4": true,
    "weapon_healer_0": true, "weapon_healer_1": true, "weapon_healer_2": true, "weapon_healer_3": true, "weapon_healer_4": true, "weapon_healer_5": true,
    "weapon_rogue_0": true, "weapon_rogue_1": true, "weapon_rogue_2": true, "weapon_rogue_3": true, "weapon_rogue_4": true, "weapon_rogue_5": true,
    "weapon_warrior_0": true, "weapon_warrior_1": true, "weapon_warrior_2": true, "weapon_warrior_3": true, "weapon_warrior_4": true, "weapon_warrior_5": true,
    "weapon_wizard_0": true, "weapon_wizard_1": true, "weapon_wizard_2": true, "weapon_wizard_3": true, "weapon_wizard_4": true, "weapon_wizard_5": true,
} }},{ multi: false });
```

```
db.users.find( {
    '_id':uuid,
}, {
    'profile.name':1,
    'achievements.ultimateGearSets':1,
    'items.gear.owned':1,
}).pretty()
```

**Comment by Alys** [comment]
migration is running

**Comment by crookedneighbor** [comment]
I'd prefer a migration script over a code change. so :+1: for doing that.

**Comment by Alys** [comment]
It still bothers me that the code will fail if the false values ever don't appear for any reason. Seems unreliable.

**Comment by gisikw** [comment]
The code as written is wrong (that is to say, it does not do what it purports to do):

https://github.com/HabitRPG/habitrpg/blob/develop/common/script/index.coffee#L1716

``` coffee
user.achievements.ultimateGearSets ?= {healer: false, wizard: false, rogue: false, warrior: false}
```

Should be (e.g. or equivalent)

```
user.achievements.ultimateGearSets ?= {}
user.achievements.ultimateGearSets.healer ?= false
user.achievements.ultimateGearSets.wizard ?= false
user.achievements.ultimateGearSets.rogue ?= false
user.achievements.ultimateGearSets.warrior ?= false
```

I agree with @Alys that this should be changed at the code level. If it doesn't get fixed, we're throwing a non-explicit constraint at the data-level, which will bite us in our rears if this becomes legacy.

**Comment by crookedneighbor** [comment]
Sounds good to me.

**Comment by gisikw** [comment]
I'll get a test case breaking and push up a fix shortly :)

**Comment by gisikw** [comment]
Addressed in #5433

**Comment by crookedneighbor** [comment]
Closed with https://github.com/HabitRPG/habitrpg/pull/5433
